deps/rabbit/src/rabbit_queue_member_eval.erl

```diff
@@ -0,0 +1,206 @@
+    Ref = erlang:send_after(NewTimeout, self(), ?EVAL_MSG),
+    {noreply, State#state{timer_ref = Ref}};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%----------------------------------------------------------------------------
+%% Internal functions
+%%----------------------------------------------------------------------------
+
+eval_quroum_queue_membership(TargetSize) ->
+    LocalLeaders = rabbit_amqqueue:list_local_leaders(),
+    ExpectedNodes = rabbit_nodes:list_members(),
+    Running = rabbit_nodes:list_running(),
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize, noop).
+
+eval_quorum_members(_ExpectedNodes, _Running, [], _TargetSize, Result) ->
+    Result;
+eval_quorum_members(ExpectedNodes, Running, [Q | LocalLeaders], TargetSize, OldResult) ->
+    {ok, Members, {_, LeaderNode}} = ra:members(amqqueue:get_pid(Q)),
+
+    %% Check if Leader is indeed this node
+    Result = case LeaderNode =:= node() of
+                 true ->
+                     MemberNodes = [Node || {_, Node} <- Members],
+                     Remove = MemberNodes -- ExpectedNodes,
+                     case Remove of
+                         [] ->
+                             add_member(Q, Running, MemberNodes, TargetSize);
+                         _ ->
+                             remove_members(Q, Remove)
+                     end;
+                 false ->
+                     noop
+             end,
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize,
+                        update_result(OldResult, Result)).
+
+add_member(Q, Running, MemberNodes, TargetSize) ->
+    New = Running -- MemberNodes,
+    case should_add_node(MemberNodes, New, Q, TargetSize) of
+        true ->
+            %% In the future, sort the list of new nodes based on load,
+            %% availability zones etc
+            Node = hd(New),
+            QName = amqqueue:get_name(Q),
+            case rabbit_quorum_queue:add_member(Q, Node) of
+                ok ->
+                    rabbit_log:debug(
+                      "Added node ~ts as a member to queue ~ts as "
+                      "the queues target group size is not met and "
+                      "there are enough nodes in the cluster", [Node, QName]);
+                {error, Err} ->
+                    rabbit_log:warning(
+                      "~ts: failed to add member (replica) on node ~w, error: ~w",
+                      [rabbit_misc:rs(QName), Node, Err])
+            end,
+            ok;
+        false ->
+            noop
+    end.
+
+should_add_node(MemberNodes, New, Q, TargetSize) ->
+    CurrentSize = length(MemberNodes),
+    TargetSize = get_target_size(Q, TargetSize),
+    maybe
+        true ?= CurrentSize < TargetSize, %% Target size not reached
+        true ?= (TargetSize - CurrentSize) =< length(New), %% Enough nodes,
```

Comment:
Should this be `length(New) > 0`? As long as there is a node, we can add a new member?

Reply:
We decided to only grow if there is a chance to reach the target.

Reply:
Is it to prevent the queue from having an even number of members?